File: DataStructure/AVLTree.py
class AVLTree:

    class _Node:
        __slots__ = ['lson', 'rson', 'cnt', 'size', 'val', 'height']

        def __init__(self, val, lson=None, rson=None, cnt=1, size=1, height=1):
            self.lson = lson
            self.rson = rson
            self.cnt = cnt
            self.size = size
            self.val = val
            self.height = height

        def update(self):
            self.size = 0
            if self.lson is not None:
                self.size += self.lson.size
            if self.rson is not None:
                self.size += self.rson.size
            self.size += self.cnt
            self.height = max(self.lson.height if self.lson is not None else 0, self.rson.height if self.rson is not None else 0) + 1

    def __init__(self):
        self.root = None

    def get_val(self, p):
        if p is None:
            return None
        else:
            return p.val

    def get_size(self, p):
        if p is None:
            return 0
        else:
            return p.size

    def get_cnt(self, p):
        if p is None:
            return 0
        else:
            return p.cnt

    def get_height(self, p):
        if p is None:
            return 0
        else:
            return p.height

    def lrotate(self, node):
        tmp = node.rson
        node.rson = tmp.lson
        tmp.lson = node
        node.update()
        tmp.update()
        return tmp

    def rrotate(self, node):
        tmp = node.lson
        node.lson = tmp.rson
        tmp.rson = node
        node.update()
        tmp.update()
        return tmp

    def _add(self, node, key):
        # print(node)
        if node is None:
            node = self._Node(key)
            # print(node)
            return node
        if key < node.val:
            node.lson = self._add(node.lson, key)
            node.update()
            if self.get_height(node.lson) - self.get_height(node.rson) >= 2:
                if key <= node.lson.val:
                    node = self.rrotate(node)
                    node.update()
                else:
                    node.lson = self.lrotate(node.lson)
                    node = self.rrotate(node)
                    node.update()
        elif key > node.val:
            node.rson = self._add(node.rson, key)
            node.update()
            # print(node.rson, key)
            if self.get_height(node.rson) - self.get_height(node.lson) >= 2:
                if key >= node.rson.val:
                    node = self.lrotate(node)
                    node.update()
                else:
                    node.rson = self.rrotate(node.rson)
                    node = self.lrotate(node)
                    node.update()
        else:
            self.cnt += 1
        return node

    def add(self, key):
        if self.root is None:
            self.root = self._Node(key)
        else:
            self.root = self._add(self.root, key)


def Travel(node):
    if node == None:
        return
    print(node.val, end=' ')
    Travel(node.lson)
    Travel(node.rson)

if __name__ == '__main__':
    t = AVLTree()
    d = input().strip().split();
    l = [int(x) for x in d]
    for i in l:
        t.add(i)
        # Travel(t.root)
        # print(t.get_height(t.root.lson), t.get_height(t.root.rson))
        # print()
        # print(t.root.val)
    Travel(t.root)
    #print(t.root.lson.val)

File: BalanceTree/AVLTree.c
#include <stdio.h>
#include <stdlib.h>

#define INF 2147483647

//#define LOCAL_TEST
#define MORE_FEATURE

struct node {
    struct node *lson;
    struct node *rson;
    int val;
    int cnt;
    int size;
    int height;
};

struct node *root;

int _max(int a, int b) {return a>b?a:b;}
int size(struct node *o) {return o==NULL?0:o->size;}
int cnt(struct node *o) {return o==NULL?0:o->cnt;}
int height(struct node *o) {return o==NULL?0:o->height;}
void pushup(struct node **o) {
    (*o)->height = _max(height((*o)->lson), height((*o)->rson)) + 1;
    (*o)->size = cnt(*o) + size((*o)->lson) + size((*o)->rson);
}

void lrotate(struct node **o) {
    struct node *tmp = (*o)->rson;
    (*o)->rson = tmp->lson;
    tmp->lson = (*o);
    pushup(o); pushup(&tmp);
    (*o) = tmp;
}
void rrotate(struct node **o) {
    struct node *tmp = (*o)->lson;
    (*o)->lson = tmp->rson;
    tmp->rson = (*o);
    pushup(o); pushup(&tmp);
    (*o) = tmp;
}

void LL(struct node **o) {rrotate(o);}
void LR(struct node **o) {lrotate(&((*o)->lson)); rrotate(o);}
void RL(struct node **o) {rrotate(&((*o)->rson)); lrotate(o);}
void RR(struct node **o) {lrotate(o);}

struct node *avltree_insert(struct node *o, int key) {
    if(o == NULL) {
        o = (struct node *)malloc(sizeof(struct node));
        o->lson = o->rson = NULL;
        o->height = o->size = o->cnt = 1;
        o->val = key;
        return o;
    }
    if(key == o->val) {
        o->cnt++;
        pushup(&o);
        return o;
    } else if(key < o->val) {
        o->lson = avltree_insert(o->lson, key);
        if(height(o->lson) - height(o->rson) == 2) {
            if(key <= o->lson->val) LL(&o);
            else LR(&o);
        }
        pushup(&o);
        return o;
    } else {
        o->rson = avltree_insert(o->rson, key);
        if(height(o->rson) - height(o->lson) == 2) {
            if(key >= o->rson->val) RR(&o);
            else RL(&o);
        }
        pushup(&o);
        return o;
    }
}

struct node *avltree_delete(struct node *o, int key) {
    if(o == NULL) return NULL;
    if(key < o->val) {
        o->lson = avltree_delete(o->lson, key);
        if(height(o->rson) - height(o->lson) == 2) {
            if(height(o->rson->rson) >= height(o->rson->lson)) RR(&o);
            else RL(&o);
        }
        pushup(&o);
        return o;
    } else if(key > o->val) {
        o->rson = avltree_delete(o->rson, key);
        if(height(o->lson) - height(o->rson) == 2) {
            if(height(o->lson->lson) >= height(o->lson->rson)) LL(&o);
            else LR(&o);
        }
        pushup(&o);
        return o;
    } else if(key == o->val) {
        if(o->cnt > 1) {
            o->cnt--;
            pushup(&o);
            return o;
        }
        if(o->lson && o->rson) {
            if(height(o->lson) >= height(o->rson)) {
                struct node *p = o->lson;
                while(p->rson) p = p->rson;
                o->val = p->val; o->cnt = p->cnt;
                p->cnt = 1;
                o->lson = avltree_delete(o->lson, p->val);
                pushup(&o);
                return o;
            } else {
                struct node *p = o->rson;
                while(p->lson) p = p->lson;
                o->val = p->val; o->cnt = p->cnt;
                p->cnt = 1;
                o->rson = avltree_delete(o->rson, p->val);
                pushup(&o);
                return o;                
            }
        } else {
            struct node *tmp = o;
            if(o->lson) o = o->lson;
            else if(o->rson) o = o->rson;
            else o = NULL;
            free(tmp);
            return o;
        }
    }
    return NULL;
}

struct node *avltree_search(struct node *o, int key) {
    if(o == NULL) return NULL;
    if(o->val == key) return o;
    else if(key < o->val) return avltree_search(o->lson, key);
    else return avltree_search(o->rson, key);
}

#ifdef MORE_FEATURE

    int _searchNext(int key, struct node *o, int ans) {
        if(o->val <= key) {
            if(o->rson == NULL) return ans;
            else return _searchNext(key, o->rson, ans);
        } else {
            if(o->lson == NULL) return o->val;
            else return _searchNext(key, o->lson, o->val);
        }
    }

    int _searchPrev(int key, struct node *o, int ans) {
        if(o->val >= key) {
            if(o->lson == NULL) return ans;
            else return _searchPrev(key, o->lson, ans);
        } else {
            if(o->rson == NULL) return o->val;
            else return _searchPrev(key, o->rson, o->val);
        }
    }

    int _kth(int k, struct node *o, int rk) {
        if(!o) return -1;
        if(rk + size(o->lson) + cnt(o) >= k && rk + size(o->lson) < k) return o->val;
        else if(rk + size(o->lson) + cnt(o) < k) return _kth(k, o->rson, rk + size(o->lson) + cnt(o));
        else return _kth(k, o->lson, rk);
    }

    int _rank(int key, struct node *o) {
        struct node *i = o;
        int rk = 0;
        while(1) {
            if(!i) return rk + 1;
            if(i->val == key) return rk + size(i->lson) + 1;
            else if(i->val < key) {
                rk += size(i->lson) + cnt(i);
                i = i->rson;
            } else {
                i = i->lson;
            }
        }
    }

#endif

#ifdef LOCAL_TEST

    void _PreOrderTravel(struct node *o) {
        if(!o) return;
        printf("%d ", o->val);
        _PreOrderTravel(o->lson);
        _PreOrderTravel(o->rson);
    }



int main() {
    int n;
    scanf("%d", &n);
    while(n--) {
        int op, x;
        struct node *q;
        scanf("%d%d", &op, &x);
        switch(op) {
            case 0:
                q = _search(root, x);
                if(!q) printf("Not Found!\n");
                else printf("%d%d%d%d\n", q->cnt, q->val, q->size, q->height);
                break;
            case 1:
                root = _Insert(root, x);
                break;
            case 2:
                root = _Delete(root, x);
                break;

#ifdef MORE_FEATURE

            case 3:
                printf("%d\n", _rank(x, root));
                break;
            case 4:
                printf("%d\n", _kth(x, root, 0));
                break;
            case 5:
                printf("%d\n", _searchPrev(x, root, -INF));
                break;
            case 6:
                printf("%d\n", _searchNext(x, root, INF));
                break;
#endif
        }

#ifdef LOCAL_TEST

        _PreOrderTravel(root);
        printf("\n");

#endif
        
    }
}

#endif